File: Data Structures & Algorithms/binary-search/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1

        while l <= r:
            # mid point
            mid = r - l // 2
            # print("mid point is: ", mid)
            if target == nums[mid]:
                return mid

            # target is in right portion
            elif target > nums[mid]:
                l = mid + 1
                
            # target is in left portion
            else:
                r = mid - 1
                
        # not found
        return -1

File: Data Structures & Algorithms/is-palindrome/submission-0.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        
        int left = 0; //start point
        int right = static_cast<int>(s.length()) - 1; //end pointer

        while (left < right) {
            while (left < right && !std::isalnum(static_cast<unsigned char>(s[left]))) {
                left++;
            }
            while (left < right && !std::isalnum(static_cast<unsigned char>(s[right]))) {
                right--;
            }

            if (std::tolower(static_cast<unsigned char>(s[left])) !=
                std::tolower(static_cast<unsigned char>(s[right]))) {
                    return false;
            }
            left++;
            right--;
        }
        return true;
    }
};
